File: desktop/src/shared/core/syntax.ts
// Renderer-agnostic port of the Android client's fail-soft syntax tokenizer
// (mobile/app/src/main/kotlin/com/silencelen/huginn/ui/Syntax.kt), plus the
// two pure dialect pickers that lived beside the Compose renderer in
// TranscriptView.kt (langForTool / resultLang) — colocated here because they
// are plain data logic that any renderer needs.
//
// A deliberately small tokenizer for the code that actually shows up in a
// Claude conversation: shell commands, diffs, and short source snippets. It is
// a lexer, not a parser — and deliberately not shiki: getting a keyword wrong
// costs a colour, never text, and the whole point is readability at a glance.
// It favours the few token classes that carry meaning (comment, string,
// number, keyword, and diff signs) over completeness.
//
// Output is spans of token kinds, not colours: the theme maps kinds to
// colours separately, and a renderer must drop (never trust) any span that
// falls outside the string.

export type Tok =
  | 'plain'
  | 'keyword'
  | 'string'
  | 'number'
  | 'comment'
  | 'function'
  | 'punct'
  | 'added'
  | 'removed'
  | 'meta'

export interface Span {
  start: number
  end: number
  tok: Tok
}

export type Dialect = 'shell' | 'diff' | 'json' | 'conf' | 'plain' | 'code'

const SHELL: ReadonlySet<string> = new Set([
  'if', 'then', 'else', 'elif', 'fi', 'for', 'while', 'do', 'done', 'case', 'esac',
  'function', 'return', 'exit', 'export', 'local', 'readonly', 'set', 'unset',
  'echo', 'cd', 'sudo', 'source', 'trap', 'shift', 'eval', 'test', 'in',
])
const C_LIKE: ReadonlySet<string> = new Set([
  'val', 'var', 'fun', 'class', 'object', 'interface', 'data', 'sealed', 'enum',
  'if', 'else', 'when', 'for', 'while', 'do', 'return', 'break', 'continue',
  'import', 'package', 'private', 'public', 'internal', 'protected', 'override',
  'suspend', 'const', 'companion', 'null', 'true', 'false', 'this', 'super',
  'try', 'catch', 'finally', 'throw', 'new', 'function', 'let', 'async', 'await',
  'def', 'elif', 'None', 'True', 'False', 'self', 'lambda', 'yield', 'with', 'as',
  'from', 'type', 'struct', 'func', 'go', 'defer', 'static', 'void', 'int', 'string',
])
const JSON_KW: ReadonlySet<string> = new Set(['true', 'false', 'null'])
const NO_KW: ReadonlySet<string> = new Set()

/** Normalizes a fence label to the dialect used for tokenizing. */
export function dialect(lang: string | null | undefined): Dialect {
  const l = lang == null ? null : lang.toLowerCase().trim()
  switch (l) {
    case 'sh':
    case 'bash':
    case 'zsh':
    case 'shell':
    case 'console':
    case 'terminal':
      return 'shell'
    case 'diff':
    case 'patch':
      return 'diff'
    case 'json':
      return 'json'
    case 'yaml':
    case 'yml':
    case 'toml':
    case 'ini':
    case 'conf':
      return 'conf'
    case null:
    case '':
      return 'plain'
    default:
      return 'code'
  }
}

export function highlight(code: string, lang: string | null | undefined): Span[] {
  switch (dialect(lang)) {
    case 'diff':
      return diff(code)
    case 'shell':
      return generic(code, SHELL, true, false)
    case 'conf':
      return generic(code, NO_KW, true, false)
    case 'json':
      return generic(code, JSON_KW, false, false)
    case 'plain':
      return []
    default:
      return generic(code, C_LIKE, true, true)
  }
}

/** What a tool's input is written in, for colouring purposes. */
export function langForTool(name: string | null | undefined): string {
  switch (name) {
    case 'Bash':
    case 'BashOutput':
      return 'shell'
    default:
      return 'plain'
  }
}

/** Tool output that looks like a diff gets diff colouring; everything else plain. */
export function resultLang(result: string): string {
  const probe = result.split('\n', 6)
  return probe.some((l) => l.startsWith('+') || l.startsWith('-') || l.startsWith('@@'))
    ? 'diff'
    : 'plain'
}

/** Whole-line classification: the only thing that matters in a diff. */
function diff(code: string): Span[] {
  const out: Span[] = []
  let i = 0
  for (const line of code.split('\n')) {
    const end = i + line.length
    let tok: Tok | null = null
    if (line.startsWith('+++') || line.startsWith('---') || line.startsWith('@@')) tok = 'meta'
    else if (line.startsWith('+')) tok = 'added'
    else if (line.startsWith('-')) tok = 'removed'
    if (tok !== null && end > i) out.push({ start: i, end, tok })
    i = end + 1
  }
  return out
}

function generic(code: string, keywords: ReadonlySet<string>, hash: boolean, slash: boolean): Span[] {
  const out: Span[] = []
  let i = 0
  const n = code.length
  while (i < n) {
    const c = code.charAt(i)
    if (hash && c === '#') {
      // Comments run to end of line.
      const end = lineEnd(code, i)
      out.push({ start: i, end, tok: 'comment' })
      i = end
    } else if (slash && c === '/' && i + 1 < n && code.charAt(i + 1) === '/') {
      const end = lineEnd(code, i)
      out.push({ start: i, end, tok: 'comment' })
      i = end
    } else if (slash && c === '/' && i + 1 < n && code.charAt(i + 1) === '*') {
      const close = code.indexOf('*/', i + 2)
      const end = close < 0 ? n : close + 2
      out.push({ start: i, end, tok: 'comment' })
      i = end
    } else if (c === '"' || c === "'" || c === '`') {
      const end = stringEnd(code, i, c)
      out.push({ start: i, end, tok: 'string' })
      i = end
    } else if (isDigit(c) && (i === 0 || !isWordChar(code.charAt(i - 1)))) {
      let j = i
      while (j < n && (isLetterOrDigit(code.charAt(j)) || code.charAt(j) === '.' || code.charAt(j) === '_')) j++
      out.push({ start: i, end: j, tok: 'number' })
      i = j
    } else if (isWordStart(c)) {
      let j = i
      while (j < n && isWordChar(code.charAt(j))) j++
      const word = code.slice(i, j)
      if (keywords.has(word)) {
        out.push({ start: i, end: j, tok: 'keyword' })
      } else if (j < n && code.charAt(j) === '(') {
        // A word directly followed by '(' is being called.
        out.push({ start: i, end: j, tok: 'function' })
      } else if (word.length > 1 && i > 0 && code.charAt(i - 1) === '-') {
        // A leading flag reads as structure in a shell command.
        out.push({ start: i - 1, end: j, tok: 'meta' })
      }
      i = j
    } else {
      i++
    }
  }
  return out
}

function lineEnd(code: string, from: number): number {
  const nl = code.indexOf('\n', from)
  return nl < 0 ? code.length : nl
}

/** Handles escapes so a `\"` inside a string does not end it. */
function stringEnd(code: string, from: number, quote: string): number {
  let j = from + 1
  while (j < code.length) {
    const ch = code.charAt(j)
    if (ch === '\\') {
      j += 2
      continue
    }
    if (ch === quote) return j + 1
    // An unterminated quote must not swallow the rest of the block.
    if (ch === '\n' && quote !== '`') return j
    j++
  }
  return code.length
}

const isDigit = (c: string): boolean => /\p{Nd}/u.test(c)
const isLetter = (c: string): boolean => /\p{L}/u.test(c)
const isLetterOrDigit = (c: string): boolean => /[\p{L}\p{Nd}]/u.test(c)
const isWordStart = (c: string): boolean => isLetter(c) || c === '_' || c === '$'
const isWordChar = (c: string): boolean => isLetterOrDigit(c) || c === '_' || c === '$'

File: mobile/core/src/commonMain/kotlin/com/silencelen/huginn/ui/Syntax.kt
package com.silencelen.huginn.ui

/**
 * A deliberately small syntax tokenizer for the code that actually shows up in a
 * Claude conversation: shell commands, diffs, and short source snippets.
 *
 * It is a lexer, not a parser. Getting a keyword wrong costs a colour, never
 * correctness, and the whole point is readability at a glance on a phone — so it
 * favours the few token classes that carry meaning (comment, string, number,
 * keyword, and diff signs) over completeness.
 *
 * Kept free of Compose types so it can be unit-tested on the JVM.
 */
object Syntax {

    enum class Tok { PLAIN, KEYWORD, STRING, NUMBER, COMMENT, FUNCTION, PUNCT, ADDED, REMOVED, META }

    data class Span(val start: Int, val end: Int, val tok: Tok)

    private val SHELL = setOf(
        "if", "then", "else", "elif", "fi", "for", "while", "do", "done", "case", "esac",
        "function", "return", "exit", "export", "local", "readonly", "set", "unset",
        "echo", "cd", "sudo", "source", "trap", "shift", "eval", "test", "in",
    )
    private val C_LIKE = setOf(
        "val", "var", "fun", "class", "object", "interface", "data", "sealed", "enum",
        "if", "else", "when", "for", "while", "do", "return", "break", "continue",
        "import", "package", "private", "public", "internal", "protected", "override",
        "suspend", "const", "companion", "null", "true", "false", "this", "super",
        "try", "catch", "finally", "throw", "new", "function", "let", "async", "await",
        "def", "elif", "None", "True", "False", "self", "lambda", "yield", "with", "as",
        "from", "type", "struct", "func", "go", "defer", "static", "void", "int", "string",
    )

    /** Normalizes a fence label to the dialect used for tokenizing. */
    fun dialect(lang: String?): String = when (lang?.lowercase()?.trim()) {
        "sh", "bash", "zsh", "shell", "console", "terminal" -> "shell"
        "diff", "patch" -> "diff"
        "json" -> "json"
        "yaml", "yml", "toml", "ini", "conf" -> "conf"
        null, "" -> "plain"
        else -> "code"
    }

    fun highlight(code: String, lang: String?): List<Span> = when (dialect(lang)) {
        "diff" -> diff(code)
        "shell" -> generic(code, SHELL, hash = true, slash = false)
        "conf" -> generic(code, emptySet(), hash = true, slash = false)
        "json" -> generic(code, setOf("true", "false", "null"), hash = false, slash = false)
        "plain" -> emptyList()
        else -> generic(code, C_LIKE, hash = true, slash = true)
    }

    /** Whole-line classification: the only thing that matters in a diff. */
    private fun diff(code: String): List<Span> {
        val out = mutableListOf<Span>()
        var i = 0
        for (line in code.split("\n")) {
            val end = i + line.length
            val tok = when {
                line.startsWith("+++") || line.startsWith("---") || line.startsWith("@@") -> Tok.META
                line.startsWith("+") -> Tok.ADDED
                line.startsWith("-") -> Tok.REMOVED
                else -> null
            }
            if (tok != null && end > i) out.add(Span(i, end, tok))
            i = end + 1
        }
        return out
    }

    private fun generic(code: String, keywords: Set<String>, hash: Boolean, slash: Boolean): List<Span> {
        val out = mutableListOf<Span>()
        var i = 0
        val n = code.length
        while (i < n) {
            val c = code[i]
            when {
                // Comments run to end of line.
                hash && c == '#' -> {
                    val end = lineEnd(code, i); out.add(Span(i, end, Tok.COMMENT)); i = end
                }
                slash && c == '/' && i + 1 < n && code[i + 1] == '/' -> {
                    val end = lineEnd(code, i); out.add(Span(i, end, Tok.COMMENT)); i = end
                }
                slash && c == '/' && i + 1 < n && code[i + 1] == '*' -> {
                    val close = code.indexOf("*/", i + 2)
                    val end = if (close < 0) n else close + 2
                    out.add(Span(i, end, Tok.COMMENT)); i = end
                }
                c == '"' || c == '\'' || c == '`' -> {
                    val end = stringEnd(code, i, c)
                    out.add(Span(i, end, Tok.STRING)); i = end
                }
                c.isDigit() && (i == 0 || !isWordChar(code[i - 1])) -> {
                    var j = i
                    while (j < n && (code[j].isLetterOrDigit() || code[j] == '.' || code[j] == '_')) j++
                    out.add(Span(i, j, Tok.NUMBER)); i = j
                }
                isWordStart(c) -> {
                    var j = i
                    while (j < n && isWordChar(code[j])) j++
                    val word = code.substring(i, j)
                    when {
                        word in keywords -> out.add(Span(i, j, Tok.KEYWORD))
                        // A word directly followed by '(' is being called.
                        j < n && code[j] == '(' -> out.add(Span(i, j, Tok.FUNCTION))
                        // A leading flag reads as structure in a shell command.
                        word.length > 1 && i > 0 && code[i - 1] == '-' -> out.add(Span(i - 1, j, Tok.META))
                    }
                    i = j
                }
                else -> i++
            }
        }
        return out
    }

    private fun lineEnd(code: String, from: Int): Int {
        val nl = code.indexOf('\n', from)
        return if (nl < 0) code.length else nl
    }

    /** Handles escapes so a `\"` inside a string does not end it. */
    private fun stringEnd(code: String, from: Int, quote: Char): Int {
        var j = from + 1
        while (j < code.length) {
            val ch = code[j]
            if (ch == '\\') { j += 2; continue }
            if (ch == quote) return j + 1
            // An unterminated quote must not swallow the rest of the block.
            if (ch == '\n' && quote != '`') return j
            j++
        }
        return code.length
    }

    private fun isWordStart(c: Char) = c.isLetter() || c == '_' || c == '$'
    private fun isWordChar(c: Char) = c.isLetterOrDigit() || c == '_' || c == '$'
}
